parser/prism/Translator.cc

```diff
@@ -890,6 +891,139 @@ unique_ptr<parser::Node> Translator::translate(pm_node_t *node, bool preserveCon
+
+                    // The parser::Send case makes a fake parser::Array with locZeroLen to hide callWithSplat
+                    // methods from hover. Using the array's loc means that we will get a zero-length loc for
+                    // the Splat in that case, and non-zero if there was a real Array literal.
+                    // if (auto splattedExpr = ast::MK::extractSplattedExpression(expr)) {
+                    //     // Exception::raise("hello, world!");
+                    //     // Extract the argument from the old Send and create a new one with array's location
+                    //     expr = MK::Splat(location.copyWithZeroLength(), move(splattedExpr));
+                    // }
+
+                    ENFORCE(expr != nullptr);
+                    argExprs.emplace_back(move(expr));
+                }
+                auto argsArrayExpr = desugarArray(location.copyWithZeroLength(), prismArgs, move(argExprs));
+
+                // Build up an array that represents the keyword args for the send.
+                // When there is a Kwsplat, treat all keyword arguments as a single argument.
+                // If the kwargs hash is not present, make a `nil` to put in the place of that argument.
+                // This will be used in the implementation of the intrinsic to tell the difference between keyword
+                // args, keyword args with kw splats, and no keyword args at all.
+                ExpressionPtr kwargsExpr;
+                if (kwargsHash != nullptr) {
+                    ast::Array::ENTRY_store kwargElements;
+                    flattenKwargs(kwargsHash, kwargElements);
+                    ast::desugar::DuplicateHashKeyCheck::checkSendArgs(ctx, 0, kwargElements);
+
+                    // Add the kwargs Hash back into parse tree, so that it's correct, too.
+                    // This doesn't effect the desugared expression.
+                    args.emplace_back(move(kwargsHash));
+
+                    kwargsExpr = MK::Array(loc, move(kwargElements));
+                } else {
+                    kwargsExpr = MK::Nil(loc);
+                }
+
+                if (prismBlock && PM_NODE_TYPE_P(prismBlock, PM_BLOCK_ARGUMENT_NODE)) {
+                    // Add the parser node back into the wq tree, to pass the parser tests.
+                    args.emplace_back(move(blockPassNode));
+                }
+
+                auto numPosArgs = 4;
+                ast::Send::ARGS_store magicSendArgs;
+                magicSendArgs.reserve(numPosArgs); // TODO: reserve room for a block pass arg
+                magicSendArgs.emplace_back(move(receiverExpr));
+                magicSendArgs.emplace_back(move(methodName));
+                magicSendArgs.emplace_back(move(argsArrayExpr));
+                magicSendArgs.emplace_back(move(kwargsExpr));
+
+                if (prismBlock != nullptr && PM_NODE_TYPE_P(prismBlock, PM_BLOCK_ARGUMENT_NODE) &&
```

Comment:
This block is analogous to this this case:

https://github.com/sorbet/sorbet/blob/13459ea30ab9e3ac2ead5ca4411549ddd9940131/ast/desugar/Desugar.cc#L872-L879